jenkins/worker.py

```diff
@@ -0,0 +1,223 @@
+USER_TIME_PATTERN = Pattern(r'^\s*User time \(seconds\): (?P<user_time>\d+.\d+|\d+)')
+SYSTEM_TIME_PATTERN = Pattern(r'^\s*System time \(seconds\): (?P<system_time>\d+.\d+|\d+)')
+WALL_TIME_PATTERN = Pattern(r'^\s*Elapsed \(wall clock\) time \(h:mm:ss or m:ss\): ((?P<hours>\d{1,2}):|)'
+                            r'(?P<minutes>\d{1,2}):(?P<seconds>\d{2}\.\d{2})')
+MAX_RES_SET_SIZE_PATTERN = Pattern(r'^\s*Maximum resident set size \(kbytes\): (?P<max_set_size>\d+)')
+
+
+# Pattern matcher for extracting the HEAD commit SHA-1 hash.
+GIT_SHA1_PATTERN = Pattern(r'^(?P<sha1>[0-9a-f]{40})\s+HEAD')
+
+
+# Returns the SHA-1 hash of the commit tagged HEAD.
+def get_tot_git_sha1():
+    output = subprocess.check_output(['git', 'ls-remote', 'https://github.com/googlecartographer/cartographer.git'])
+    dict = GIT_SHA1_PATTERN.extract(output)
+    return dict['sha1']
+
+
+# Returns a dictionary of stats.
+def extract_stats(input):
+    result = {}
+
+    dict = USER_TIME_PATTERN.extract(input)
+    result['user_time_secs'] = float(dict['user_time'])
+
+    dict = SYSTEM_TIME_PATTERN.extract(input)
+    result['system_time_secs'] = float(dict['system_time'])
+
+    dict = WALL_TIME_PATTERN.extract(input)
+    wall_time_secs = 0.0
+    if (dict['hours'] != None):
+        wall_time_secs += int(dict['hours']) * 3600
+    wall_time_secs += int(dict['minutes']) * 60
+    wall_time_secs += float(dict['seconds'])
+    result['wall_time_secs'] = wall_time_secs
+
+    dict = MAX_RES_SET_SIZE_PATTERN.extract(input)
+    result['max_set_size_kbytes'] = int(dict['max_set_size'])
+
+    return result
+
+
+# Convenience function for Datastore key construction and entity retrieval.
+def retrieve_entity(datastore_client, kind, id):
+    key = datastore_client.key(kind, id)
+    return datastore_client.get(key)
+
+
+# Constructs a round-robin job selector.
+def create_job_selector(worker_id, num_workers):
+    return lambda job_id: job_id % num_workers == worker_id
+
+
+# Runs the provided command both printing its stdout output to stdout and returning the output as a string.
+def run_cmd(cmd):
+    p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    output = ''
+    while p.poll() is None:
+        l = p.stdout.readline()
+        output += l + '\n'
+        print l
+
+    l = p.stdout.read()
+    output += l
+    print l
+    return output
+
+
+# Represents a single job to be executed. A job consists of a combination of rosbag and configuration and launch files.
+class Job:
+    def __init__(self, datastore_client, job_id):
+        self.id = job_id
+        entity = retrieve_entity(datastore_client, "Job", job_id)
+        self.launch_file = entity['launch_file']
+        self.assets_writer_launch_file = entity['assets_writer_launch_file']
+        self.assets_writer_config_file = entity['assets_writer_config_file']
+        self.rosbag = entity['rosbag']
+
+    def __str__(self):
```

Comment:
should this not be `__repr__`?  Also, add the class name into this

Reply:
Done